File: backend/TestCaseGenerator.cs
using System;
using System.Collections.Generic;
using backend.DTO;
using TestingBackend;

namespace backend
{
    public static class TestCaseGenerator
    {
        public static HashSet<List<IInput>> GenerateTestCases(List<IInput> inputs)
        {
            var output = new HashSet<List<IInput>>();

            output.Add(CalculateInOnPatterns1(inputs));
            output.Add(CalculateInOnPatterns2(inputs));

            var oof = OffOut(inputs);

            foreach(var xd in oof)
            {
                output.Add(xd);
            }

            return output;
        }

        private static List<IInput> CalculateInOnPatterns1(List<IInput> inputs)
        {
            List<IInput> output = new();

            foreach (var input in inputs)
            {
                switch (input.Expression)
                {
                    case Expressions.Lower:
                    case Expressions.LowerAndEqual:
                        output.Add(In(input, 1));
                    break;

                    case Expressions.Higher:
                    case Expressions.HigherAndEqual:
                    case Expressions.NotEqual:
                        output.Add(In(input, 2));
                    break;

                    case Expressions.Equal:
                        output.Add(On(input));
                    break;

                    case Expressions.BoolTrue:
                    case Expressions.BoolFalse:
                        output.Add(input);
                    break;

                    case Expressions.Interval:
                        output.Add(On(input, 2));
                    break;
                    
                    default:
                    break;
                }
            }

            return output;
        }

        private static List<IInput> CalculateInOnPatterns2(List<IInput> inputs)
        {
            List<IInput> output = new();

            foreach (var input in inputs)
            {
                switch (input.Expression)
                {
                    case Expressions.Lower:
                    case Expressions.LowerAndEqual:
                        output.Add(On(input, 1));
                    break;
                    case Expressions.Equal:
                        output.Add(On(input));
                    break;

                    case Expressions.Higher:
                    case Expressions.HigherAndEqual:
                        output.Add(On(input, 2));
                    break;

                    case Expressions.NotEqual:
                        output.Add(In(input, 1));
                    break;

                    case Expressions.BoolTrue:
                    case Expressions.BoolFalse:
                        output.Add(input);
                    break;

                    case Expressions.Interval:
                        output.Add(On(input, 1));
                    break;
                    
                    default:
                    break;
                }
            }

            return output;
        }

        private static List<IInput> BaseLine(List<IInput> inputs)
        {
            List<IInput> output = new();

            foreach (var input in inputs)
            {
                switch (input.Expression)
                {
                    case Expressions.Lower:
                    case Expressions.LowerAndEqual:
                        output.Add(In(input, 1));
                    break;

                    case Expressions.Higher:
                    case Expressions.HigherAndEqual:
                    case Expressions.NotEqual:
                        output.Add(In(input, 2));
                    break;

                    case Expressions.Equal:
                        output.Add(On(input));
                    break;

                    case Expressions.BoolTrue:
                    case Expressions.BoolFalse:
                        output.Add(input);
                    break;

                    case Expressions.Interval:
                        output.Add(In(input, 3));
                    break;
                    
                    default:
                    break;
                }
            }

            return output;
        }

        private static HashSet<List<IInput>> OffOut(List<IInput> inputs)
        {
            var output = new HashSet<List<IInput>>();

            for (var i = 0; i < inputs.Count; ++i)
            {
                var based1 = BaseLine(inputs);
                var based2 = BaseLine(inputs);

                switch (inputs[i].Expression)
                {
                    case Expressions.Lower:
                    case Expressions.LowerAndEqual:
                        based1[i] = Out(inputs[i], 1);
                        based2[i] = Off(inputs[i], 1);

                        output.Add(based1);
                        output.Add(based2);
                    break;
                    case Expressions.Higher:
                    case Expressions.HigherAndEqual:
                        based1[i] = Out(inputs[i], 2);
                        based2[i] = Off(inputs[i], 2);

                        output.Add(based1);
                        output.Add(based2);
                    break;

                    case Expressions.Equal:
                        based1[i] = Out(inputs[i], 1);
                        based2[i] = Out(inputs[i], 2);

                        output.Add(based1);
                        output.Add(based2);
                    break;

                    case Expressions.NotEqual:
                        based1[i] = Off(inputs[i]);

                        output.Add(based1);
                    break;

                    case Expressions.BoolTrue:
                        based1[i] = new BoolDTO(inputs[i].Expression, false);
                        output.Add(based1);
                    break;
                    case Expressions.BoolFalse:
                        based1[i] = new BoolDTO(inputs[i].Expression, true);
                        output.Add(based1);
                    break;

                    case Expressions.Interval:
                        var based3 = BaseLine(inputs);
                        var based4 = BaseLine(inputs);

                        based1[i] = Out(inputs[i], 1);
                        based2[i] = Out(inputs[i], 2);
                        based3[i] = Off(inputs[i], 1);
                        based4[i] = Off(inputs[i], 2);

                        output.Add(based1);
                        output.Add(based4);
                        output.Add(based3);
                        output.Add(based2);
                    break;
                }
            }

            return output;
        }

        private static IInput On(IInput input, int version = 0)
        {
            var temp = input as IntervalDTO;
            switch (version)
            {
                case 1 : // <, <=, Interval Right
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.NumToNum(
                            (temp.Interval.IntervalData.High - (temp.Interval.IsOpen.High ? 1 : 0) * temp.Precision, 
                            temp.Interval.IntervalData.High - (temp.Interval.IsOpen.High ? 1 : 0) * temp.Precision),
                            (false, false)
                            ),
                            temp.Precision
                        );
                case 2 : // >, >=, Interval left
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.NumToNum(
                            (temp.Interval.IntervalData.Low + (temp.Interval.IsOpen.Low ? 1 : 0) * temp.Precision,
                            temp.Interval.IntervalData.Low + (temp.Interval.IsOpen.Low ? 1 : 0) * temp.Precision),
                            (false, false)
                            ),
                            temp.Precision
                        );
                default : // ==
                    return new IntervalDTO(input.Expression, temp.Interval, temp.Precision);
            }
        }

        private static IInput? In(IInput input, int version = 0)
        {
            var temp = input as IntervalDTO;
            switch (version)
            {
                case 1 : // <, <=
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.InfToNum(
                            temp.Interval.IntervalData.High - (temp.Interval.IsOpen.High ? 2 : 1) * temp.Precision,
                            (true, false)
                            ),
                            temp.Precision
                        );
                case 2 : // >, =>
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.NumToInf(
                            temp.Interval.IntervalData.Low + (temp.Interval.IsOpen.Low ? 2 : 1) * temp.Precision,
                            (false, true)
                            ),
                            temp.Precision
                        );
                case 3 : // Interval
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.NumToNum(
                            (temp.Interval.IntervalData.Low + (temp.Interval.IsOpen.Low ? 2 : 1) * temp.Precision,
                            temp.Interval.IntervalData.High - (temp.Interval.IsOpen.High ? 2 : 1) * temp.Precision),
                            (false, false)
                            ),
                            temp.Precision
                        );
                default :
                    return null;
            }
        }

        private static IInput? Off(IInput input, int version = 0)
        {
            var temp = input as IntervalDTO;
            switch (version)
            {
                case 1 : // <, <=, Interval Right, == right
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.NumToNum(
                            (temp.Interval.IntervalData.High + (temp.Interval.IsOpen.High ? 0 : 1) * temp.Precision, 
                            temp.Interval.IntervalData.High + (temp.Interval.IsOpen.High ? 0 : 1) * temp.Precision),
                            (false, false)
                            ),
                            temp.Precision
                        );
                case 2 : // >, >=, // Interval left, == left
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.NumToNum(
                            (temp.Interval.IntervalData.Low - (temp.Interval.IsOpen.Low ? 0 : 1) * temp.Precision,
                            temp.Interval.IntervalData.Low - (temp.Interval.IsOpen.Low ? 0 : 1) * temp.Precision),
                            (false, false)
                            ),
                            temp.Precision
                        );
                default :
                    return new IntervalDTO(
                        temp.Expression,
                        temp.Interval,
                        temp.Precision
                    );
            }
        }

        private static IInput? Out(IInput input, int version = 0)
        {
            var temp = input as IntervalDTO;
            switch (version)
            {
                case 1 : // <, <=, Interval Right
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.NumToInf(
                            temp.Interval.IntervalData.High + (temp.Interval.IsOpen.High ? 1 : 2) * temp.Precision,
                            (false, true)
                            ),
                            temp.Precision
                        );
                case 2 : // >, =>, Interval Left
                    return new IntervalDTO(
                        temp.Expression,
                        Interval.InfToNum(
                            temp.Interval.IntervalData.Low - (temp.Interval.IsOpen.Low ? 1 : 2) * temp.Precision,
                            (true, false)
                            ),
                            temp.Precision
                        );
                default :
                    return null;
            }
        }
    }
}

File: src/logic/testCaseGenerator.ts
import { array } from 'fp-ts';
import { Interval } from 'interval-arithmetic';
import {
  BoolDTO,
  Expression,
  IInput,
  IntervalDTO,
  MissingVariableDTO,
} from './models/dtos';
import { NTuple } from './models/ntuple';

export function generateTestCases(inputs: IInput[]): NTuple[] {
  const nTuples = [
    calculateInOnPatterns1(inputs),
    calculateInOnPatterns2(inputs),
    ...OffOut(inputs),
  ].map((x) => new NTuple(x));

  return array.uniq(NTuple.Eq)(nTuples);
}

const calculateInOnPatterns1 = (inputs: IInput[]): IInput[] =>
  // eslint-disable-next-line array-callback-return
  inputs.map((input) => {
    if (input.isConstant) return input;

    switch (input.expression) {
      case Expression.LessThan:
      case Expression.LessThanOrEqualTo:
        return InIn(input, 1);

      case Expression.GreaterThan:
      case Expression.GreaterThanOrEqualTo:
        return InIn(input, 2);

      case Expression.NotEqualTo:
        return In(input, 2);

      case Expression.EqualTo:
        return On(input);

      case Expression.BoolTrue:
      case Expression.BoolFalse:
      case Expression.MissingVariable:
        return input;

      case Expression.Interval:
        return On(input, 2);
    }
  });

const calculateInOnPatterns2 = (inputs: IInput[]): IInput[] =>
  // eslint-disable-next-line array-callback-return
  inputs.map((input) => {
    if (input.isConstant) return input;

    switch (input.expression) {
      case Expression.LessThan:
      case Expression.LessThanOrEqualTo:
        return On(input, 1);

      case Expression.EqualTo:
        return On(input);

      case Expression.GreaterThan:
      case Expression.GreaterThanOrEqualTo:
        return On(input, 2);

      case Expression.NotEqualTo:
        return In(input, 1);

      case Expression.BoolTrue:
      case Expression.BoolFalse:
      case Expression.MissingVariable:
        return input;

      case Expression.Interval:
        return On(input, 1);
    }
  });

const baseline = (inputs: IInput[]): IInput[] =>
  // eslint-disable-next-line array-callback-return
  inputs.map((input) => {
    if (input.isConstant) return input;

    switch (input.expression) {
      case Expression.LessThan:
      case Expression.LessThanOrEqualTo:
        return In(input, 1);

      case Expression.GreaterThan:
      case Expression.GreaterThanOrEqualTo:
      case Expression.NotEqualTo:
        return In(input, 2);

      case Expression.EqualTo:
        return On(input);

      case Expression.BoolTrue:
      case Expression.BoolFalse:
      case Expression.MissingVariable:
        return input;

      case Expression.Interval:
        return In(input, 3);
    }
  });

function OffOut(inputs: IInput[]): IInput[][] {
  const output = [];

  for (let i = 0; i < inputs.length; ++i) {
    if (inputs[i] instanceof MissingVariableDTO) continue;
    if (inputs[i].isConstant) continue;

    var based1 = baseline(inputs);
    var based2 = baseline(inputs);

    switch (inputs[i].expression) {
      case Expression.LessThan:
      case Expression.LessThanOrEqualTo:
        based1[i] = Out(inputs[i], 1);
        based2[i] = Off(inputs[i], 1);

        output.push(based1);
        output.push(based2);
        break;

      case Expression.GreaterThan:
      case Expression.GreaterThanOrEqualTo:
        based1[i] = Out(inputs[i], 2);
        based2[i] = Off(inputs[i], 2);

        output.push(based1);
        output.push(based2);
        break;

      case Expression.EqualTo:
        based1[i] = Out(inputs[i], 3);
        based2[i] = Out(inputs[i], 4);

        output.push(based1);
        output.push(based2);
        break;

      case Expression.NotEqualTo:
        based1[i] = Off(inputs[i]);

        output.push(based1);
        break;

      case Expression.BoolTrue:
        based1[i] = new BoolDTO(inputs[i].expression, false);
        output.push(based1);
        break;

      case Expression.BoolFalse:
        based1[i] = new BoolDTO(inputs[i].expression, true);
        output.push(based1);
        break;

      case Expression.Interval:
        var based3 = baseline(inputs);
        var based4 = baseline(inputs);

        based1[i] = Out(inputs[i], 1);
        based2[i] = Out(inputs[i], 2);
        based3[i] = Off(inputs[i], 1);
        based4[i] = Off(inputs[i], 2);

        output.push(based1);
        output.push(based4);
        output.push(based3);
        output.push(based2);
        break;
    }
  }

  return output;
}

function On(input: IInput, version: 0 | 1 | 2 = 0): IInput {
  if (!(input instanceof IntervalDTO))
    throw new Error("On's argument must be an IntervalDTO");

  switch (version) {
    // <, <=, Interval Right
    case 1:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.hi - (input.isOpen.hi ? 1 : 0) * input.precision,
          input.interval.hi - (input.isOpen.hi ? 1 : 0) * input.precision,
        ),
        input.precision,
      );
    // >, >=, Interval left
    case 2:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.lo + (input.isOpen.lo ? 1 : 0) * input.precision,
          input.interval.lo + (input.isOpen.lo ? 1 : 0) * input.precision,
        ),
        input.precision,
      );
    // ==
    case 0:
      return new IntervalDTO(input.expression, input.interval, input.precision);
  }
}

function In(input: IInput, version: 1 | 2 | 3): IInput {
  if (!(input instanceof IntervalDTO))
    throw new Error("In's argument must be an IntervalDTO");

  switch (version) {
    // <, <=
    case 1:
      return new IntervalDTO(
        input.expression,
        new Interval(
          -Infinity,
          input.interval.hi - (input.isOpen.hi ? 1 : 0) * input.precision,
        ),
        input.precision,
      );
    // >, =>
    case 2:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.lo + (input.isOpen.lo ? 1 : 0) * input.precision,
          Infinity,
        ),
        input.precision,
      );
    // Interval
    case 3:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.lo + (input.isOpen.lo ? 1 : 0) * input.precision,
          input.interval.hi - (input.isOpen.hi ? 1 : 0) * input.precision,
        ),
        input.precision,
      );
  }
}

function InIn(input: IInput, version: 1 | 2): IInput {
  if (!(input instanceof IntervalDTO))
    throw new Error("InIn's argument must be an IntervalDTO");

  switch (version) {
    // <, <=
    case 1:
      return new IntervalDTO(
        input.expression,
        new Interval(
          -Infinity,
          input.interval.hi - (input.isOpen.hi ? 2 : 1) * input.precision,
        ),
        input.precision,
      );
    // >, =>
    case 2:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.lo + (input.isOpen.lo ? 2 : 1) * input.precision,
          Infinity,
        ),
        input.precision,
      );
  }
}

function Off(input: IInput, version: number = 0): IInput {
  if (!(input instanceof IntervalDTO))
    throw new Error("Off's argument must be an IntervalDTO");

  switch (version) {
    // <, <=, Interval Right, == right
    case 1:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.hi + (input.isOpen.hi ? 0 : 1) * input.precision,
        ),
        input.precision,
      );
    // >, >=, // Interval left, == left
    case 2:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.lo - (input.isOpen.lo ? 0 : 1) * input.precision,
        ),
        input.precision,
      );

    default:
      return new IntervalDTO(input.expression, input.interval, input.precision);
  }
}

function Out(input: IInput, version: 1 | 2 | 3 | 4): IInput {
  if (!(input instanceof IntervalDTO))
    throw new Error("Off's argument must be an IntervalDTO");

  switch (version) {
    // <, <=, Interval Right
    case 1:
      return new IntervalDTO(
        input.expression,
        new Interval(
          input.interval.hi + (input.isOpen.hi ? 1 : 2) * input.precision,
          Infinity,
        ),
        input.precision,
      );

    // >, =>, Interval Left
    case 2:
      return new IntervalDTO(
        input.expression,
        new Interval(
          -Infinity,
          input.interval.lo - (input.isOpen.lo ? 1 : 2) * input.precision,
        ),
        input.precision,
      );
    // =, Right
    case 3:
      return new IntervalDTO(
        input.expression,
        new Interval(input.interval.hi + input.precision, Infinity),
        input.precision,
      );
    // =, Left
    case 4:
      return new IntervalDTO(
        input.expression,
        new Interval(-Infinity, input.interval.lo - input.precision),
        input.precision,
      );
  }
}
